ArduPlane/stall_detection.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2020  Kraus Hamdani Aerospace Inc. All rights reserved.
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+ * Author: Tom Pittenger, Samuel Tabor
+ */
+
+
+#include "Plane.h"
+
+/*
+  stall detection algorithm and anti-stall management
+ */
+
+
+const AP_Param::GroupInfo Plane::StallState::var_info[] = {
+    // @Param: STALL_DETECT
+    // @DisplayName: Stall detection type
+    // @Description: Stall detection type. Different aircraft stall in different ways.
+    // @User: Advanced
+    // @Bitmask: 0:CantHoldAltitude,1:SinkRate*2Max,2:SinkRate*4Max,3:RollError20deg,4:RollError30deg,5:RollError45deg,6:PitchError10deg,7:PitchError20deg,8:PitchError20deg,9:PitchError30deg,10:PitchError40deg,11:AltError10m,12:AltError20m,13:AltError40m,14:AltError60m
+    AP_GROUPINFO("DETECT", 0, StallState, detection_bitmask, 0),
+
+    AP_GROUPEND
+};
+
+
+/*
+  Do we think we are stalled?
+  Probabilistic method where a bool is low-passed and considered a probability.
+*/
+void Plane::StallState::detection_update(void)
+{
+    if (!plane.arming.is_armed() ||
+        !plane.is_flying() ||
+        plane.crash_state.is_crashed)
+    {
+        stall_clear();
+        count = 0;
+        detection_log();
+        return;
+    }
+
+    const bool is_stalled_initial = is_stalled();
+
+    last_detection = detect_stall(true);
+
+    // LPF confidence
+    confidence = ((float)last_detection * LPFcoef) + (confidence * (1.0f - LPFcoef));
```

Comment:
`last_detection` is a bool, casting to a float then multiplying just seems very wrong.